src/parsers/archaea_parsers.py

```diff
@@ -0,0 +1,692 @@
+from pyspark.sql.types import ArrayType, StringType, StructType, StructField
+
+## XML namespace mapping for UniProt entries (used for all XPath queries)
+NS = {"u": "https://uniprot.org/uniprot"}
+
+
+def generate_cdm_id(accession:str) -> str:
+    """ 
+    Generate a deterministic CDM entity_id based on a UniProt accession.
+    Uses MD5 hash to ensure fixed length and avoid exposing raw accession.
+    """
+    if not accession or not isinstance(accession, str):
+        raise ValueError("accession must be a non-empty string")
+    normalized = accession.strip()
+    md5_hash = hashlib.md5(normalized.encode("utf-8")).hexdigest()
+    return f"CDM:{md5_hash}"
+
+
+def extract_version(url):
+    """
+    Extracts UniProt version string from a download URL
+    """
+    patterns = [(r'release[-_]?(\d{4}_\d{2})', 1)]
+    for pat, group in patterns:
+        match_obj = re.search(pat, url, re.IGNORECASE)
+        if match_obj:
+            return match_obj.group(group)
+    return None
+   
+
+def build_datasource_record(xml_url):
+    """
+    Build a provenance record for the UniProt datasource
+    """
+    name = "UniProt import"
+    version = extract_version(xml_url) if xml_url else None
+    accessed = int(datetime.datetime.now(datetime.timezone.utc).timestamp())
+    return {
+        "name": name,
+        "source": "UniProt",
+        "url": xml_url,
+        "accessed": accessed,
+        "version": version,
+    }
+
+
+def parse_identifiers(entry, cdm_id):
+    """
+    Extract all accession numbers in the UniProt entry and format them into a CDM identifier structure
+    """
+    return [
+        {
+            "entity_id": cdm_id,
+            "identifier": f"UniProt:{acc.text}",
+            "source": "UniProt",
+            "description": "UniProt accession"
+        }
+        for acc in entry.findall("u:accession", NS)
+    ]
+
+
+def parse_names(entry, cdm_id):
+    """
+    Extract all protein names from a UniProt <entry> element, including
+    - Top-level <name> elements (generic names)
+    - <recommendedName> and <alternativeName> blocks within <protein> (full and short names)
+    """
+    names = []
+
+    # Extract all top-level <name> tags 
+    for name_element in entry.findall("u:name", NS):
+        if name_element.text:
+            names.append({
+                "entity_id": cdm_id,
+                "name": name_element.text,
+                "description": "UniProt protein name",
+                "source": "UniProt"
+            })
+
+    # Extract recommended and alternative names from <protein> block
+    protein = entry.find("u:protein", NS)
+    if protein is not None:
+        for name_type in ["recommended", "alternative"]:
+            # recommendedName is singular, alternativeName can have multiple
+            if name_type == "recommended":
+                name_blocks = [protein.find("u:recommendedName", NS)]
+            else:
+                name_blocks = protein.findall("u:alternativeName", NS)
```

Comment:
You can simplify the code here by writing this as
```py
  name_blocks = protein.findall(f"u:{name_type}Name", NS)
```

If there's only ever one recommended name, `protein.findall(...)` will always return a list with only one entry. The if/else is not needed.